File: formacOS/Sources/AnyClipCore/Hashing.swift
import CryptoKit
import Foundation

/// Hex SHA-256 of raw bytes — same as Python's hashlib.sha256(data).hexdigest().
public func sha256Hex(_ data: Data) -> String {
    SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
}

/// Hex SHA-256 of the UTF-8 encoding of `text`.
public func sha256Hex(_ text: String) -> String {
    sha256Hex(Data(text.utf8))
}

/// Echo-suppression key for a multi-file clip. Sort the per-file sha256
/// lowercase-hex strings lexicographically (hex is ASCII, so Swift's default
/// String `<` gives the required plain ordinal order), concatenate with no
/// separator, and sha256 the ASCII bytes. Order-independent so pasteboard
/// re-detection order can never break suppression. Keep in lockstep with
/// anyclip.aggregate_files_hash and C# Hashing.AggregateFilesHash.
public func aggregateFilesHash(_ hashes: [String]) -> String {
    sha256Hex(hashes.sorted().joined())
}

File: forwindows/src/AnyClipCore/Hashing.cs
using System.Security.Cryptography;
using System.Text;

namespace AnyClip.Core;

public static class Hashing
{
    public static string Sha256Hex(byte[] data) =>
        Convert.ToHexString(SHA256.HashData(data)).ToLowerInvariant();

    public static string Sha256Hex(string text) =>
        Sha256Hex(Encoding.UTF8.GetBytes(text));

    /// Echo-suppression key for a multi-file clip: sort per-file sha256 hex
    /// strings by ordinal, concatenate with no separator, sha256 the bytes.
    /// Order-independent. Keep in lockstep with Swift/Python.
    public static string AggregateFilesHash(IEnumerable<string> hashes) =>
        Sha256Hex(string.Concat(hashes.OrderBy(h => h, StringComparer.Ordinal)));
}
